File: sets/sets.js
class Tuple {
    constructor(x, y, z) {
        this.x = x;
        this.y = y;
        this.z = z;
    }

    product() {
        return this.x * this.y * this.z;
    }

    toString() {
        return `${this.x}, ${this.y}, ${this.z}`
    }
}

function timeIt(callback, repetitions) {
    const start = performance.now();
    for (let i = 0; i < repetitions; i++) {
        callback(i);
    }
    const end = performance.now();
    return end - start;
}

function perfRun(callback, repetitions, trials, name, init=null) {
    console.log('Run:', name)
    for(let i=0; i<trials; ++i) {
        if(init!=null) init(i);
        const time = timeIt(callback, repetitions);
        console.log(`Trial ${i} took ${time}`);
    }
    console.log();
}

const mySet = new Set();
const mySetStr = new Set();
const myArray = [];
const count = 100000;
const uniques = 1000;
const tuples = []
for(let i=0;i<count;i++) {
    tuples.push(new Tuple(i%uniques+4,i%uniques-7,i%uniques+100));
}

// Add objects to the Set
perfRun((i) => {
    const tuple = tuples[i]
    mySet.add(tuple);
}, count, 5, `Add ${count} objects with ${uniques} values to empty Set`, (trial)=>mySet.clear());
console.log('Set size', mySet.size);
console.log();
console.log();

// Add objects to the String Set
perfRun((i) => {
    const tuple = tuples[i]
    mySetStr.add(tuple.toString());
}, count, 5, `Add ${count} objects with ${uniques} values  to empty String Set`, (trial)=>mySetStr.clear());
console.log('String set size', mySetStr.size);
console.log();
console.log();


// Add objects to the Array
perfRun((i) => {
    const tuple = tuples[i].toString();
    if(myArray.indexOf(tuple)<0) {
        myArray.push(tuple);
    }
}, count, 5, `Add ${count} objects with ${uniques} values to empty Array implementation of set`, (trial)=>myArray.splice(0,myArray.length));
console.log('Array size', myArray.length);
console.log();
console.log();

// Lookup objects in the Set
perfRun((i) => {
    const tuple = tuples[i];
    return mySet.has(tuple);
}, count, 5, `Time to lookup ${count} objects in Set`);

// Lookup objects in the String Set
perfRun((i) => {
    const tuple = tuples[i];
    return mySetStr.has(tuple.toString());
}, count, 5, `Time to lookup ${count} objects in String Set`);

// Lookup objects in the Array
perfRun((i) => {
    const tuple = tuples[i].toString();
    return myArray.indexOf(tuple)>=0;
}, count, 5, `Time to lookup ${count} objects in Array implementation of Set`);

// Remove objects from the Set
perfRun((i) => {
    const tuple = tuples[i];
    mySet.delete(tuple);
}, count, 1, `Time to remove ${count} objects from Set`);
// console.log(`Set size: ${mySet.size}`);
// console.log();
// console.log();

// Remove objects from the Set
perfRun((i) => {
    const tuple = tuples[i];
    mySetStr.delete(tuple.toString());
}, count, 1, `Time to remove ${count} objects from String Set`);
// console.log(`String set size: ${mySetStr.size}`);
// console.log();
// console.log();

// Remove objects from the Set
perfRun((i) => {
    const tuple = tuples[i].toString();
    const ind = myArray.indexOf(tuple);
    if(ind>=0) myArray.splice(ind,1);
}, count, 1, `Time to remove ${count} objects from Array implementation of Set`);
// console.log(`Array size: ${myArray.length}`);
// console.log();
// console.log();

File: sets/sets.py
import time

class Tuple:
    def __init__(self, x, y, z):
        self.x = x
        self.y = y
        self.z = z

    def product(self):
        return self.x * self.y * self.z

    def __str__(self):
        return f"{self.x}, {self.y}, {self.z}"

    # Ensure that the Tuple can be used as a dictionary key or in a set
    def __hash__(self):
        return hash((self.x, self.y, self.z))

    def __eq__(self, other):
        return (self.x, self.y, self.z) == (other.x, other.y, other.z)

def time_it(callback, repetitions):
    start = time.perf_counter()
    for i in range(repetitions):
        callback(i)
    end = time.perf_counter()
    return (end - start) * 1000  # Convert to milliseconds

def perf_run(callback, repetitions, trials, name, init=None):
    print('Run:', name)
    for i in range(trials):
        if init is not None:
            init(i)
        time_taken = time_it(callback, repetitions)
        print(f'Trial {i} took {time_taken:.2f}ms')
    print()

count = 100000
uniques = 1000
tuples = [Tuple(i % uniques + 4, i % uniques - 7, i % uniques + 100) for i in range(count)]

# Using set for unique Tuple objects
my_set = set()
perf_run(lambda i: my_set.add(tuples[i]), count, 5, f'Add {count} objects with {uniques} values to empty Set', lambda trial: my_set.clear())
print('Set size', len(my_set))
print()

# Using set for unique string representations of Tuple objects
my_set_str = set()
perf_run(lambda i: my_set_str.add(str(tuples[i])), count, 5, f'Add {count} objects with {uniques} values to empty String Set', lambda trial: my_set_str.clear())
print('String set size', len(my_set_str))
print()

# Using list to keep unique string representations of Tuple objects
my_list = []
perf_run(lambda i: my_list.append(str(tuples[i])) if str(tuples[i]) not in my_list else None, count, 5, f'Add {count} objects with {uniques} values to empty List implementation of Set', lambda trial: my_list.clear())
print('List size', len(my_list))
print()

# Lookup objects in the set
perf_run(lambda i: tuples[i] in my_set, count, 5, 'Time to lookup objects in Set')

# Lookup string representations in the string set
perf_run(lambda i: str(tuples[i]) in my_set_str, count, 5, 'Time to lookup objects in String Set')

# Lookup string representations in the list
perf_run(lambda i: str(tuples[i]) in my_list, count, 5, 'Time to lookup objects in List')

# Remove objects from the set
perf_run(lambda i: my_set.discard(tuples[i]), count, 1, 'Time to remove objects from Set')
# print('Set size:', len(my_set))
# print()

# Remove string representations from the string set
perf_run(lambda i: my_set_str.discard(str(tuples[i])), count, 1, 'Time to remove objects from String Set')
# print('String set size:', len(my_set_str))
# print()

# Remove string representations from the list
perf_run(lambda i: my_list.remove(str(tuples[i])) if str(tuples[i]) in my_list else None, count, 1, 'Time to remove objects from List')
# print('List size:', len(my_list))
# print()
